### ISSUE TITLE
UE5 (Unreal Engine 5 Game Engine)

### ISSUE THREAD

**Issue by Ashe** [NixOS/nixpkgs#124963]
**Project description**
Now that we have access to Unreal Engine 5's source code, it would be nice to have a ue5 package which corresponds to the main branch of UE5 `ue5-main`. 

The [quick start](https://docs.unrealengine.com/4.26/en-US/SharingAndReleasing/Linux/BeginnerLinuxDeveloper/SettingUpAnUnrealWorkflow/) documentation shows some steps to getting it working, however as a NixOS newbie I got stuck when it begun requiring things of `dotnet` (I also wasn't making a package which was just making my life more difficult). 

Apparently the stages to get this setup are:
> 1. Clone the branch
> 2. Run `Setup.sh ` in the root folder
> 3. Run `GenerateProjectFiles.sh` in the root folder
> 4. Run `make`

Some things to note:
* Their scripts use `/bin/bash` which need changing to be `/bin/sh` (or maybe there's a better way of doing this with Nix)
* There are internal programs for `mono` and `dotnet` (I think), however I couldn't get these working. Trying to use my own versions of these sort of worked but never fully got things working

Would love to see this! Thank you!

**Metadata**

* homepage URL:https://www.unrealengine.com/en-US/unreal-engine-5?sessionInvalidated=true
* source URL: https://github.com/EpicGames/UnrealEngine/tree/ue5-main
* license: Unreal® Engine End User License Agreement
* platforms: Linux

**Comment by prusnak** [comment]
> Now that we have access to Unreal Engine 5's source code

Who is "we"? The repo https://github.com/EpicGames/UnrealEngine is not public

**Comment by Ashe** [comment]
> > Now that we have access to Unreal Engine 5's source code
> 
> Who is "we"? The repo https://github.com/EpicGames/UnrealEngine is not public

You can get access to UE source code by connecting your GitHub with them. I don't like it, but it is open to anybody who wants it. I presume it's what we do with the UE4 package:

https://github.com/NixOS/nixpkgs/blob/nixos-20.09/pkgs/games/ue4/default.nix#L76

`url = "https://github.com/EpicGames/UnrealEngine/releases/tag/${version}";` - same repo, just on the UE5 branch rather than ue4.

**Comment by juliosueiras** [comment]
I am taking a jab at ue5 right now, there is quite a few changes compare to UE4 like introduction of bundled tooling(which require some modification to bypass), I am hoping within today/tomorrow I will have a working version of it (since it also require nuget packages as well now)

**Comment by juliosueiras** [comment]
some progress update, dividing the process to execute GitDeps as its own derivation, due to the massive unpacking it does

**Comment by juliosueiras** [comment]
so, the current big issue with both UE5 and latest UE4 is that everything is self-bundled(in a hardcoded fashion), currently trying buildFHSUserEnv approach(for now to get it building), will most likely have more progress mid june once getting new equipment, due to the high process power and storage space that UE5 need

**Comment by juliosueiras** [comment]
(to give perspective, I split the process of GitDeps and that result in 42GB of deps)

**Comment by IvarWithoutBones** [comment]
> Their scripts use /bin/bash which need changing to be /bin/sh (or maybe there's a better way of doing this with Nix)

Just wanted to add that the `patchShebangs` function can do this, see for example: https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/pkgs/applications/office/bookworm/default.nix#L57

**Comment by juliosueiras** [comment]
will be posting progress by the end of the week

**Comment by juliosueiras** [comment]
![Screenshot from 2021-06-19 05-24-54](https://user-images.githubusercontent.com/3680302/122638049-d7a24900-d0bf-11eb-893a-4dc1111d2df3.png)
![Screenshot from 2021-06-19 05-26-58](https://user-images.githubusercontent.com/3680302/122638050-d8d37600-d0bf-11eb-86c7-91f4f3cd192b.png)

the current progress, pretty much done for first passthrough, this is via nix's clang not the bundled clang, going spend the next week to convert to pure drv

**Comment by Ashe** [comment]
Absolutely fantastic work, cannot wait to have this on my machine!

**Comment by juliosueiras** [comment]
initial basic check list:

* Compiling C++ or modifying the engine etc
* Plugins
* Exporting a game to linux , and possible mobile/other
* Some new UE5 features like world partitioning
* Simple multiplayer (PIE and standalone)
* Importing models and textures etc
* Controller support

**Comment by loucass003** [comment]
Any updates on this ?

**Comment by juliosueiras** [comment]
@loucass003 working on it, the only main component left is the compilation process issue for C++ based projects and exporting, but aside from that, everything else is working fine so far(blueprint, animation, etc) I am just taking a few days to not get burnt out

**Comment by juliosueiras** [comment]
will be resume working on it tomorrow(was busy with work the past week)

**Comment by loucass003** [comment]
Still nothing ? :cry:

**Comment by juliosueirasED** [comment]
> Still nothing ? :cry: 

Still working on it , mostly the current issue is to fix the hardcore path and have a full start to finish build both a C++ project(both GUI and via make) and a Blueprint project

**Comment by juliosueiras** [comment]
> Still nothing ? 😢

Still working on it , mostly the current issue is to fix the hardcore path and have a full start to finish build both a C++ project(both GUI and via make) and a Blueprint project

**Comment by loucass003** [comment]
> > Still nothing ? 😢
> 
> Still working on it , mostly the current issue is to fix the hardcore path and have a full start to finish build both a C++ project(both GUI and via make) and a Blueprint project

Ok cool, thanks for your work on it!

**Comment by juliosueiras** [comment]
will see if I can start creating the draft PR this weekend

**Comment by Ashe** [comment]
> will see if I can start creating the draft PR this weekend

Thank you for your hard work, you must have put a lot of effort into this!

**Comment by dyaso** [comment]
Any hope whatever's working so far could be published, even without an official PR?

**Comment by juliosueiras** [comment]
@dyaso yea, was mostly waiting for UE5 to be more stable, and will be resuming on it since now have access to a high-memory server farm for building it(since UE5 packaging is not exactly a easy one desktop situation)

**Comment by dyaso** [comment]
Fantastic! Yeah compiling 4.26 was bad enough, had to clear a load of space and leave my machine at it for a day; having to keep redoing the process plus whatever increased resources UE5 needs to build must be prohibitive. Thanks for making the effort.

Unfortunately I want to use Blueprint functions with the Control Rig plugin, which are only available in 5EA currently, so have to keep rebooting into Windows which is annoying for what's so far just a hobby project.

**Comment by iblech** [comment]
Just wanted to thank you @juliosueiras for your hard work, very much appreciated :-)

**Comment by PAI5REECHO** [comment]
Make a draft PR, I'll be happy to contribute build system fixes

**Comment by jerith666** [comment]
I poked around under his github account and found https://github.com/juliosueiras-nix/nix-ue5.  It just has one commit from 2021-06-23, but might be something of a starting point.

**Comment by juliosueiras** [comment]
@jerith666 will push the new update, the current hurdle is mostly boil down to way UE package its games and its cmake lookup,etc (the engine itself can run and do stuff with no problem)

**Comment by juliosueiras** [comment]
will be creating the draft PR tonight/tomorrow

**Comment by juliosueiras** [comment]
![](https://cdn.discordapp.com/attachments/546545180476440576/967528544131502141/Screenshot_from_2022-04-23_16-52-37.png)

**Comment by juliosueiras** [comment]
![Screenshot from 2022-04-23 16-57-27](https://user-images.githubusercontent.com/3680302/164945870-778e2e52-5ef4-4753-af2b-b8096ffc0806.png)
![Screenshot from 2022-04-23 16-59-56](https://user-images.githubusercontent.com/3680302/164945974-3a3bbd78-6e6a-40ab-9b83-bfada35335ba.png)


![Screenshot from 2022-04-23 17-01-17](https://user-images.githubusercontent.com/3680302/164945880-39fdd537-32ba-47dd-a442-971bb99d4450.png)

**Comment by juliosueiras** [comment]
![](https://cdn.discordapp.com/attachments/860017489311694848/967864592812371968/Screenshot_from_2022-04-24_15-07-54.png)

C++ based project was able to open(needed to tweak a few things)

**Comment by juliosueiras** [comment]
![Screenshot from 2022-04-24 18-06-32](https://user-images.githubusercontent.com/3680302/164999486-d60883b9-ea2d-4a68-8100-757a46f4b894.png)
![Screenshot from 2022-04-24 17-45-20](https://user-images.githubusercontent.com/3680302/164999490-524537b8-591c-481d-bfce-287ce6db8418.png)
Quick Launch work now(needed a few tweak) first pic is Blueprint based project second pic is C++ based project

will be postponing the submit to tomorrow, since I am downloading the City sample to test an actual production project load just to make sure before writing the draft PR
![Screenshot from 2022-04-24 18-34-22](https://user-images.githubusercontent.com/3680302/164999538-816f447d-e00d-4feb-933d-8dbda735a5eb.png)

**Comment by juliosueiras** [comment]
FYI is annoying that I need to use Epic Game Launcher(wine) to download marketplace asset/project

**Comment by juliosueiras** [comment]
Nevermind, doing the Content Showcase project instead, since City Sample is killing my bandwidth(in term of throttle)

**Comment by juliosueiras** [comment]
k, tested all aspect of the Content showcase, no severe issue(the only crash was trying to play the Niagara advanced portion, but that is most likely due to my GPU not being strong enough to handle all of it at once)

**Comment by juliosueiras** [comment]
the draft PR is here if anybody want to try it out(need to add the wrapper and thats it), keep in mind the result size is around 122GB(as usual for Unreal Engine)

**Comment by dyaso** [comment]
Trying to install [62eeb15105](https://github.com/juliosueiras/nixpkgs/tree/62eeb151055c8241e00c8d504590def07b5385b5) keeps failing for me with `error: Too many open files` . This is the end of the output:

```
  /nix/store/zz53xrgqi4z0r2g9sgq5y9xhxdyvb61w-629161f1442cc070f84ca259e983259288e25a85.drv
  /nix/store/zz7k2dvk4dacixsgz899v4gsd93rdi0s-4b04e4b958292f22dda67f4b10168f4465a1e5da.drv
  /nix/store/zz9n51hqiysywrwmnvnl4ybjgki5mip5-f7a44be87cd8b2cc22a5c6580998902cd597ffac.drv
  /nix/store/xsah6clnmk6qkhgq4mgndbv9amrvhj9s-link-deps.sh.drv
  /nix/store/djqgp32g9nppwdlw1c65vq9qm8w9akas-ue5-unwrapped-5.0.1.drv
  /nix/store/7p3hypzpy1263fqwk3wp9sm8lldb8r4d-run-ue5.drv
  /nix/store/9l5jyhh2l365741h5m9q1gcf44vzvrzx-NVIDIA-Linux-x86_64-510.39.01.run.drv
  /nix/store/q2jw4zpp4zqxm6s65fpgs587vvkhpb07-nvidia-x11-510.39.01-5.15.34.drv
  /nix/store/b5szdf0zsrv6p2jsdxfhs95rfg8lhn47-UnrealEditor-usr-multi.drv
  /nix/store/dpnbwavz2a0gxr144lzwqxi0sy868fj3-UnrealEditor-init.drv
  /nix/store/lp5bvwgirryki7fl8an28b7x9pj5b4s9-ldconfig.drv
  /nix/store/m7mi8qrqpmkwr1j1c4lnhjd4drf3n3yx-profile.drv
  /nix/store/n9pfh8f189czc072dgnd9px7jigsy8hz-UnrealEditor-chrootenv-etc.drv
  /nix/store/gqfz9vfrcydhiaa393w6945bl27w5jmj-UnrealEditor-usr-target.drv
  /nix/store/m6v27rp04nwf2rnc5m94b2r4y27knf40-UnrealEditor-fhs.drv
  /nix/store/lzrsh8kg3pi79zzh31r73q660c2yk28y-UnrealEditor.drv
  /nix/store/bcn9i9rf31y36ynwlnmvnxsj72vcjr6d-UnrealEditor.drv
building '/nix/store/jahqlpa0mrnxzp43drlb2qvgdzmlkw1d-builder.pl.drv'...
building '/nix/store/lp5bvwgirryki7fl8an28b7x9pj5b4s9-ldconfig.drv'...
building '/nix/store/m7mi8qrqpmkwr1j1c4lnhjd4drf3n3yx-profile.drv'...
error: opening directory '/nix/store/pjwr9vhy4npycfgifhn5vx0xj24xwv7f-b4cc1896317297900c5e91fe25eb2f0c7a2cea8c': Too many open files
```

**Comment by juliosueiras** [comment]
@dyaso I think you might need to set the ulimit for nix-daemon itself, for ex. I have this for my machine `systemd.services.nix-daemon.serviceConfig.LimitNOFILE = pkgs.lib.mkForce "40096";` , since is hardcoded by default to 1024

**Comment by dyaso** [comment]
That worked, thanks! Will try building again this evening. Huge gratitude for your work on this

**Comment by juliosueiras** [comment]
@dyaso fyi, the purpose of the wrapper is for tweaking and figuring out which packages is needed it for runtime, since for example, there are a few files path that need to be copied instead of symlink for Unreal to work(for building a game)

**Comment by juliosueiras** [comment]
but at the same time, this tweak won't cause rebuild for the base package, so no need to wait for 122GB rebuild

**Comment by dyaso** [comment]
I'm most interested in using the level and animation editors, and the in-editor python scripting (which doesn't run in the game engine, just automates tasks within the Editor itself). I don't need to build and package games separately, so long as the Play In Editor feature works (the play button on the top central toolbar). If you have the content examples working, that is everything I was hoping for.

Unfortunately, now I'm getting an error about not being able to download `UnrealEngine-5.0.1-release.zip` automatically, despite just downloading and adding it to the store as advised.

<details><summary>Tail of the installation output</summary>

```
$ nix-store --add-fixed sha256 downloads/UnrealEngine-5.0.1-release.zip 
/nix/store/asnk5fc2wv6br0qcasa0rbm6vf3biy0x-UnrealEngine-5.0.1-release.zip
```

Btw i'm trying to install the whole thing with `$ nix-env -f ue5nix/ -iA 'ue5'` , hope that's correct.

```
/nix/store/zz7k2dvk4dacixsgz899v4gsd93rdi0s-4b04e4b958292f22dda67f4b10168f4465a1e5da.drv
  /nix/store/zz9n51hqiysywrwmnvnl4ybjgki5mip5-f7a44be87cd8b2cc22a5c6580998902cd597ffac.drv
  /nix/store/xsah6clnmk6qkhgq4mgndbv9amrvhj9s-link-deps.sh.drv
  /nix/store/djqgp32g9nppwdlw1c65vq9qm8w9akas-ue5-unwrapped-5.0.1.drv
  /nix/store/7p3hypzpy1263fqwk3wp9sm8lldb8r4d-run-ue5.drv
  /nix/store/9l5jyhh2l365741h5m9q1gcf44vzvrzx-NVIDIA-Linux-x86_64-510.39.01.run.drv
  /nix/store/q2jw4zpp4zqxm6s65fpgs587vvkhpb07-nvidia-x11-510.39.01-5.15.34.drv
  /nix/store/b5szdf0zsrv6p2jsdxfhs95rfg8lhn47-UnrealEditor-usr-multi.drv
  /nix/store/dpnbwavz2a0gxr144lzwqxi0sy868fj3-UnrealEditor-init.drv
  /nix/store/gqfz9vfrcydhiaa393w6945bl27w5jmj-UnrealEditor-usr-target.drv
  /nix/store/m6v27rp04nwf2rnc5m94b2r4y27knf40-UnrealEditor-fhs.drv
  /nix/store/lzrsh8kg3pi79zzh31r73q660c2yk28y-UnrealEditor.drv
  /nix/store/bcn9i9rf31y36ynwlnmvnxsj72vcjr6d-UnrealEditor.drv
building '/nix/store/lpa799dhcbj11gbgr7gjpp6ywwcsr2hr-UnrealEngine-5.0.1-release.zip.drv'...
building '/nix/store/rdic62lf2fb07afhfp68gxpby06j51r5-06ff1948527143f9d1793ee543d2dcfb747a855f.drv'...
building '/nix/store/bf14nnfk7n7mhpwglj1bbxl7gpvz02f1-070808461825eec7222431e9b3ea81fb7e43b80e.drv'...
building '/nix/store/6z36i4cgqg6rk3nkiy34jyhs8rpbx07s-089f18ba43366ca33d2d15e83d1858473d9c2d61.drv'...

***
Unfortunately, we cannot download file UnrealEngine-5.0.1-release.zip automatically.
Please go to https://github.com/EpicGames/UnrealEngine/releases/tag/5.0.1-release to download it yourself, and add it to the Nix store
using either
  nix-store --add-fixed sha256 UnrealEngine-5.0.1-release.zip
or
  nix-prefetch-url --type sha256 file:///path/to/UnrealEngine-5.0.1-release.zip

***

error: builder for '/nix/store/lpa799dhcbj11gbgr7gjpp6ywwcsr2hr-UnrealEngine-5.0.1-release.zip.drv' failed with exit code 1;
       last 10 log lines:
       > ***
       > Unfortunately, we cannot download file UnrealEngine-5.0.1-release.zip automatically.
       > Please go to https://github.com/EpicGames/UnrealEngine/releases/tag/5.0.1-release to download it yourself, and add it to the Nix store
       > using either
       >   nix-store --add-fixed sha256 UnrealEngine-5.0.1-release.zip
       > or
       >   nix-prefetch-url --type sha256 file:///path/to/UnrealEngine-5.0.1-release.zip
       >
       > ***
       >
       For full logs, run 'nix log /nix/store/lpa799dhcbj11gbgr7gjpp6ywwcsr2hr-UnrealEngine-5.0.1-release.zip.drv'.
error: 1 dependencies of derivation '/nix/store/djqgp32g9nppwdlw1c65vq9qm8w9akas-ue5-unwrapped-5.0.1.drv' failed to build
building '/nix/store/9l5jyhh2l365741h5m9q1gcf44vzvrzx-NVIDIA-Linux-x86_64-510.39.01.run.drv'...
error: 1 dependencies of derivation '/nix/store/7p3hypzpy1263fqwk3wp9sm8lldb8r4d-run-ue5.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dpnbwavz2a0gxr144lzwqxi0sy868fj3-UnrealEditor-init.drv' failed to build
error: 1 dependencies of derivation '/nix/store/lzrsh8kg3pi79zzh31r73q660c2yk28y-UnrealEditor.drv' failed to build
error: 1 dependencies of derivation '/nix/store/bcn9i9rf31y36ynwlnmvnxsj72vcjr6d-UnrealEditor.drv' failed to build
```

</details>

**Comment by juliosueiras** [comment]
@dyaso what is the sha256sum of the zip?

**Comment by juliosueiras** [comment]
my zip is `09545b0cf2c435e27838f806ad9e0f3e7d7e48f744816c1e99b4a4ed10a659ec`

**Comment by dyaso** [comment]
`1c296e6f3f8d1eff960053b23ce95e115645c2d899a8c3f7810813d91f135e18`, both times i've dowloaded it today

**Comment by juliosueiras** [comment]
@dyaso thanks, found the issue, is due to the method of download from my end(the inside name of the zip is using hash instead of the tag which cause the sha256 to be different), will push the correct hash in a bit

**Comment by dyaso** [comment]
So close!

First build attempt, starting with ~182 GB free, took between 9-11 hours downloading and compiling before getting an out-of-space error and 124 GB free -- presumably it deletes partial compilation results unless the whole thing's successful?

Cleared another ~25 GB, tried again; at one point saw it sitting still with the message "buildPhase completed in 5 hours 13 minutes 22 seconds" and 24 GB free, came back an hour later to same disk space errer, attached below. Again, 124GB are now free.

It talks at the top about an Andried physics library -- as an example, is there any way to disable that component? I don't need to package things, just to use the Editor.

Is there a way to avoid the 5 hours rebuild time each attempt, if only the installation step is failing? Or maybe i could run it from where it's built, without adding it to the store?

Otherwise, now attempting to clear a further 100 GB...

<details><summary>Final 124 lines of build output, from where the errors start appearing</summary>

```
[287/290] Link (lld) UnrealInsights
[288/290] Link (lld) libUnrealInsights-GameplayInsights.so
[289/290] WriteMetadata UnrealInsights.version
[290/290] WriteMetadata UnrealInsights.target
Total time in Parallel executor: 778.88 seconds
Total execution time: 782.26 seconds
buildPhase completed in 5 hours 13 minutes 22 seconds
installing
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3ExtensionsPROFILE.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3ExtensionsPROFILE.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPxPvdSDKCHECKED.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPxPvdSDKCHECKED.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CharacterKinematicCHECKED.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CharacterKinematicCHECKED.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CharacterKinematicPROFILE.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CharacterKinematicPROFILE.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CharacterKinematic.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CharacterKinematic.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3VehiclePROFILE.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3VehiclePROFILE.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPsFastXml.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPsFastXml.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPsFastXmlCHECKED.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPsFastXmlCHECKED.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPsFastXmlPROFILE.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPsFastXmlPROFILE.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3ExtensionsCHECKED.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3ExtensionsCHECKED.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CookingCHECKED.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CookingCHECKED.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3VehicleCHECKED.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3VehicleCHECKED.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPxFoundationPROFILE.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPxFoundationPROFILE.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3Cooking.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3Cooking.a': No space left on device
mv: error copying 'Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CHECKED.a' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x64/libPhysX3CHECKED.a': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/x86': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Android/ARM64': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Mac': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/HoloLens': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/TVOS': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Win64': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/IOS': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/PhysX3/Lib/Unix': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Protobuf': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Python3': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Qualcomm': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Re2': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/RenderDoc': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/SDL2': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/SPIRV-Reflect': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/ShaderConductor': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/SpeedTree': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Steamworks': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Tencent': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Unix': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/VHACD': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Vorbis': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Vulkan': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/WebRTC': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/WiX': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/WinHttp': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/Windows': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/WindowsMixedRealityInterop': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/asio': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/coremod': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/dlmalloc': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/fastJSON': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/glslang': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/heapprofd': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/hlslcc': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/jemalloc': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libJPG': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libOpus': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libPNG': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libPhonon': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libSampleRate': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libWebSockets': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libcurl': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libjpeg-turbo': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libstrophe': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/libxml2': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/llvm': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/metis': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/mimalloc': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/mtlpp': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/nanosvg': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/nvTextureTools': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/nvTriStrip': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/nvtesslib': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/openexr': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/portmidi': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/rpclib': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/vcpkg': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/xxhash': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/zlib': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/ThirdParty/HoloLens': No space left on device
mv: error copying 'Engine/Source/UnrealClient.Target.cs' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/UnrealClient.Target.cs': No space left on device
mv: error copying 'Engine/Source/UnrealEditor.Target.cs' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/UnrealEditor.Target.cs': No space left on device
mv: error copying 'Engine/Source/UnrealGame.Target.cs' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/UnrealGame.Target.cs': No space left on device
mv: error copying 'Engine/Source/UnrealServer.Target.cs' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Source/UnrealServer.Target.cs': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Content': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Saved': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Engine/Intermediate': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/FeaturePacks': No space left on device
mv: error copying 'GenerateProjectFiles.bat' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/GenerateProjectFiles.bat': No space left on device
mv: error copying 'GenerateProjectFiles.command' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/GenerateProjectFiles.command': No space left on device
mv: error copying 'GenerateProjectFiles.sh' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/GenerateProjectFiles.sh': No space left on device
mv: error copying 'LICENSE.md' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/LICENSE.md': No space left on device
mv: error copying 'Makefile' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Makefile': No space left on device
mv: error copying 'README.md' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/README.md': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Samples': No space left on device
mv: error copying 'Setup.bat' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Setup.bat': No space left on device
mv: error copying 'Setup.command' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Setup.command': No space left on device
mv: error copying 'Setup.sh' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Setup.sh': No space left on device
mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Templates': No space left on device
mv: error copying 'UE5.code-workspace' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/UE5.code-workspace': No space left on device
mv: error copying 'cpp.hint' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/cpp.hint': No space left on device
error: builder for '/nix/store/sj81r8ivbs5yh5nnm027gbd9sdrrxlmm-ue5-unwrapped-5.0.1.drv' failed with exit code 1;
       last 10 log lines:
       > mv: error copying 'LICENSE.md' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/LICENSE.md': No space left on device
       > mv: error copying 'Makefile' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Makefile': No space left on device
       > mv: error copying 'README.md' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/README.md': No space left on device
       > mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Samples': No space left on device
       > mv: error copying 'Setup.bat' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Setup.bat': No space left on device
       > mv: error copying 'Setup.command' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Setup.command': No space left on device
       > mv: error copying 'Setup.sh' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Setup.sh': No space left on device
       > mv: cannot create directory '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/Templates': No space left on device
       > mv: error copying 'UE5.code-workspace' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/UE5.code-workspace': No space left on device
       > mv: error copying 'cpp.hint' to '/nix/store/hswpj3ll7w5if3xmrzmci6cbw2185vri-ue5-unwrapped-5.0.1/cpp.hint': No space left on device
       For full logs, run 'nix log /nix/store/sj81r8ivbs5yh5nnm027gbd9sdrrxlmm-ue5-unwrapped-5.0.1.drv'.
       note: build failure may have been caused by lack of free disk space
error: 1 dependencies of derivation '/nix/store/0cxlrwsiz80r6bq6dm9wsh1h40qih9zd-run-ue5.drv' failed to build
error: 1 dependencies of derivation '/nix/store/w6ihbnh5r2pb7p35b00k0kk5rkm58m2h-UnrealEditor-init.drv' failed to build
error: 1 dependencies of derivation '/nix/store/kl2h16mmcjsagzj3c7s78nv2nbj5a4x2-UnrealEditor.drv' failed to build
error: 1 dependencies of derivation '/nix/store/6nw0s2jhk168jcl7qqap99hxd1v5a50h-UnrealEditor.drv' failed to build
```

</details>

**Comment by juliosueiras** [comment]
as of right now not sure(since I am adapting from what ue4 pkg had)

**Comment by juliosueiras** [comment]
try with 150/160 GB free space if you can

**Comment by dyaso** [comment]
175 GB free, am going to try `nix-build ue5nix/ -A 'ue5'` first, hopefully that'll produce a `results` symlink i can run without installing but which won't require more rebuilds from scratch? Sorry, my nix knowledge is really poor; i started reading the pills, but then resorted to booting into Windows to run things like UE and have forgotten most of it

**Comment by juliosueiras** [comment]
that would be correct A) remember is still a draft PR so additional fixes may require B) but the fixes would only be done to the wrapper level, so it won't require additional rebuild for the UE5 itself

**Comment by dyaso** [comment]
Earlier it was sitting there with the "buildPhase completed" message again, free space was 9.5GB, now "fixupPhase completed in 9 minutes 59 seconds" and 127 GB are free.

Editor is starting atm, compiling thousands of shaders

**Comment by juliosueirasED** [comment]
@dyaso was the startup crash happened before or after the project selection/creation page?

**Comment by juliosueiras** [comment]
@dyaso I pushed a new commit try that one(just incase), it won't trigger the base(120ish GB) rebuild

**Comment by dyaso** [comment]
Oh sorry, that was during the first splash screen, before project selection.

Your new version seems to've fixed it, i got to the project window and have created a Blueprints 3rd person template project, it's starting now. Will try out the content examples next

**Comment by dyaso** [comment]
Btw, just found out that if one wants to use the handy [send2unreal Blender extension](https://epicgames.github.io/BlenderTools/send2ue/), you have to enable a port for IPC by adding `networking.firewall.allowedUDPPorts = [ 6766 ];` to `configuration.nix`

**Comment by dyaso** [comment]
Just saw there's been [a hotfix release](https://github.com/EpicGames/UnrealEngine/releases/tag/5.0.2-release) with [release notes](https://forums.unrealengine.com/t/5-0-2-hotfix-released/566149) mentioning fixing "[Build Blocker] Compile UnrealHeaderTool Linux - Failed to load app-local ICU". Might that help with getting C++ games compiling?

**Comment by juliosueiras** [comment]
@dyaso yea, the error is cause because we need to remove a few specific files/directory in the linkage, to let UE rebuild again because UE like to modify itself

**Comment by juliosueiras** [comment]
will take a look at the hotfix release

**Comment by dyaso** [comment]
Managed to get it creating a C++ project without errors,

<details><summary>details inside, written as i was going</summary>

First i tried replacing the link into the store at `~/.config/unreal-engine-nix-workdir/Engine/Intermediate/Build/XmlConfigCache.bin` with a copy of the file itself, but it's still trying to write to the store copy.

Then replacing `~/.config/unreal-engine-nix-workdir/Engine/Build/BatchFiles/Linux/Build.sh` with the copy from the store, and giving the final line the argument ` -XmlConfigCache="$HOME/.config/unreal-engine-nix-workdir/Engine/Intermediate/Build/XmlConfigCache.bin"`, fixes that problem.

Now the error instead is that
<details><summary>it's trying to write a *log file* to the store</summary>

```
Running /home/user/.config/unreal-engine-nix-workdir/Engine/Build/BatchFiles/Linux/Build.sh  -projectfiles -project="/home/user/documents/Unreal Projects/MyProject/MyProject.uproject" -game -engine -progress
Running command : dotnet Engine/Binaries/DotNET/UnrealBuildTool/UnrealBuildTool.dll -projectfiles -project=/home/user/documents/Unreal Projects/MyProject/MyProject.uproject -game -engine -progress
Log file: /nix/store/9dsk0qz55b7yvfzjfjw3m46p7xdj68f4-ue5-unwrapped-5.0.2/Engine/Programs/UnrealBuildTool/Log_GPF.txt
ERROR: Unhandled exception: System.IO.IOException: Read-only file system
          at System.IO.FileSystem.LinkOrCopyFile(String sourceFullPath, String destFullPath)
          at System.IO.FileSystem.MoveFile(String sourceFullPath, String destFullPath, Boolean overwrite)
          at System.IO.File.Move(String sourceFileName, String destFileName, Boolean overwrite)
          at EpicGames.Core.Log.AddFileWriter(String Name, FileReference OutputFile) in /build/UnrealEngine-5.0.2-release/Engine/Source/Programs/Shared/EpicGames.Core/Log.cs:line 231
          at UnrealBuildTool.GenerateProjectFilesMode.Execute(CommandLineArguments Arguments) in /build/UnrealEngine-5.0.2-release/Engine/Source/Programs/UnrealBuildTool/Modes/GenerateProjectFilesMode.cs:line 72
          at UnrealBuildTool.UnrealBuildTool.Main(String[] ArgumentsArray) in /build/UnrealEngine-5.0.2-release/Engine/Source/Programs/UnrealBuildTool/UnrealBuildTool.cs:line 612
```
</details>

adding `-Log="$HOME/uelog.txt` writes a log file there, but ( [contrary to this](https://github.com/EpicGames/UnrealEngine/blob/d11782b9046e9d0b130309591e4efc57f4b8b037/Engine/Source/Programs/UnrealBuildTool/UnrealBuildTool.cs#L268) ) also still the one in the store.

(Changing `~/.config/Unreal Engine/UnrealBuildTool/BuildConfiguration.xml` doesn't appear to have any effect btw.)

Replaced `~/.config/unreal-engine-nix-workdir/Engine/Programs/UnrealBuildTool/Log_GPF.txt` with the copy from the store. No effect.

If it's still writing the default log file due to [this line](https://github.com/EpicGames/UnrealEngine/blob/d11782b9046e9d0b130309591e4efc57f4b8b037/Engine/Source/Programs/UnrealBuildTool/Modes/GenerateProjectFilesMode.cs#L71), an obvious idea was to modify and recompile UnrealBuildTool in `~/.config/unreal-engine-nix-workdir/`. However, just compiling a new copy of UBT inside `~/.config/` has got it writing to the correct Log_GPF.txt file??

On to the next errors, which concern writing to the 4 things in `.config/unreal-engine-nix-workdir/Engine/Intermediate/Build/BuildRules/` ... fixing those, and it creates a C++ project.

Trying to compile the project from within the editor, 

`ERROR: Unhandled exception: System.UnauthorizedAccessException: Access to the path '/home/user/.config/unreal-engine-nix-workdir/Engine/Intermediate/Build/SourceFileCache.bin' is denied.`

. . .
</details>

Is this similar to the "tweaks" you did, or am i going down the wrong path?

**Comment by dyaso** [comment]
@juliosueiras how did you get a C++ project to open/build?

I've been trying all weekend, so far accumulated a list of 5685 paths the Editor either tries to write to or Editor modules it says would need to be rebuilt to open the project, `~/.config/unreal-engine-nix-workdir` is now 11GB. I would quite like to get this working, even just a verbal description of what you did would be helpful

**Comment by vlinkz** [comment]
Downloading the binary from https://www.unrealengine.com/en-US/linux, and just running `DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 steam-run ./Engine/Binaries/Linux/UnrealEditor` everything seemed to work out of the box, including c++ projects and vscode!
![Screenshot from 2022-07-20 14-23-04](https://user-images.githubusercontent.com/20145996/180086108-3d9f4fe6-92da-4980-8f59-fb30973ac420.png)

**Comment by juliosueiras** [comment]
@vlinkz that would be one option(though the issue mostly would be the fact that it will be binary package, the steam-run is not much of a big issue)

**Comment by juliosueiras** [comment]
(I am fine with the binary approach, since it would be smaller and faster)

**Comment by juliosueiras** [comment]
(XD just realize, it will still be the same issue, since you are running directly, but the package will be in nix store and Unreal like to edit the engine in-place)

**Comment by frantisekhanzlikbl** [comment]
> Maybe instead packaging something like [Epic Asset Manager](https://github.com/AchetaGames/Epic-Asset-Manager) would be better. I heard the dev mention that with the new binary releases, they were planning on implementing direct downloads and version management in the app

Please note that, at the moment, Epic doesn't include Quixel Bridge in the precompiled binaries which EAM also uses, so it is still pretty important to get it compiling with nix.

**Comment by devhell** [comment]
Hey, this is awesome! I'd love to be able to use UE5 on NixOS. Has there been any progress?

**Comment by hussein-aitlahcen** [comment]
> > > Downloading the binary from https://www.unrealengine.com/en-US/linux, and just running `DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 steam-run ./Engine/Binaries/Linux/UnrealEditor` everything seemed to work out of the box, including c++ projects and vscode! ![Screenshot from 2022-07-20 14-23-04](https://user-images.githubusercontent.com/20145996/180086108-3d9f4fe6-92da-4980-8f59-fb30973ac420.png)
> > 
> > 
> > Not in my case, in my case it crashes, im trying to use 5.2, the error is the following
> > ```
> > [2023.06.11-21.22.31:155][  0]LogCore: Error: appError called: Fatal error: [File:./Runtime/VulkanRHI/Private/VulkanUtil.cpp] [Line: 1017] 
> > VulkanRHI::vkCreateQueryPool(Device->GetInstanceHandle(), &PoolCreateInfo, VULKAN_CPU_ALLOCATOR, &QueryPool) failed, VkResult=-2
> >  at ./Runtime/VulkanRHI/Private/VulkanQuery.cpp:52 
> >  with error VK_ERROR_OUT_OF_DEVICE_MEMORY
> > 0x00007f23893396e9 libUnrealEditor-VulkanRHI.so!VulkanRHI::VerifyVulkanResult(VkResult, char const*, char const*, unsigned int) [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/VulkanRHI/Private/VulkanUtil.cpp:1016]
> > 0x00007f23892e25b3 libUnrealEditor-VulkanRHI.so!FVulkanQueryPool::FVulkanQueryPool(FVulkanDevice*, FVulkanCommandBufferManager*, unsigned int, VkQueryType, bool) [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/VulkanRHI/Private/VulkanQuery.cpp:52]
> > 0x00007f23892ed95c libUnrealEditor-VulkanRHI.so!FVulkanRayTracingCompactionRequestHandler::FVulkanRayTracingCompactionRequestHandler(FVulkanDevice*) [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/VulkanRHI/Private/VulkanRayTracing.cpp:1165]
> > 0x00007f2389260891 libUnrealEditor-VulkanRHI.so!FVulkanDevice::InitGPU() [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/VulkanRHI/Private/VulkanDevice.cpp:1266]
> > 0x00007f23892fb290 libUnrealEditor-VulkanRHI.so!FVulkanDynamicRHI::InitInstance() [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/VulkanRHI/Private/VulkanRHI.cpp:751]
> > 0x00007f23892faf5d libUnrealEditor-VulkanRHI.so!FVulkanDynamicRHI::Init() [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/VulkanRHI/Private/VulkanRHI.cpp:256]
> > 0x00007f244612c26e libUnrealEditor-RHI.so!RHIInit(bool) [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/RHI/Private/DynamicRHI.cpp:347]
> > 0x0000000000253481 UnrealEditor!FEngineLoop::PreInitPreStartupScreen(char16_t const*) [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/Launch/Private/LaunchEngineLoop.cpp:3207]
> > 0x0000000000249f62 UnrealEditor!GuardedMain(char16_t const*) [/mnt/horde/++UE5/Sync/Engine/Source/./Runtime/Launch/Private/Launch.cpp:140]
> > 0x00007f24420c79ba libUnrealEditor-UnixCommonStartup.so!CommonUnixMain(int, char**, int (*)(char16_t const*), void (*)()) [/mnt/horde/++UE5/Sync/Engine/Source/Runtime/Unix/UnixCommonStartup/Private/UnixCommonStartup.cpp:269]
> > 0x00007f2441dfc24e libc.so.6!UnknownFunction(0x2924d)
> > 0x00007f2441dfc309 libc.so.6!__libc_start_main(+0x88)
> > 0x0000000000236749 UnrealEditor!_start()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > My system is good enough to run unreal engine, so i dont think is my hardware
> 
> may not be valid, my dedicated GPU was not working at that time without knowing it, so the error could have been caused by that, no clue

This works for me on NixOS with RTX 4050

**Comment by devhell** [comment]
Just tried it, doesn't seem to work with Wayland atm. :cry:

**Comment by TheOddler** [comment]
Using `DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 nvidia-offload steam-run ./Engine/Binaries/Linux/UnrealEditor` I can start the editor, create a project, and open it. But then when I press play it just crashes. I have a laptop with RTX 3050.

**Comment by HammyHavoc** [comment]
> Maybe instead packaging something like [Epic Asset Manager](https://github.com/AchetaGames/Epic-Asset-Manager) would be better. I heard the dev mention that with the new binary releases, they were planning on implementing direct downloads and version management in the app

Did this end up being the case? Is anyone successfully using UE5 on NixOS beyond just seeing if they can?

**Comment by Murazaki** [comment]
> > Maybe instead packaging something like [Epic Asset Manager](https://github.com/AchetaGames/Epic-Asset-Manager) would be better. I heard the dev mention that with the new binary releases, they were planning on implementing direct downloads and version management in the app
> 
> Did this end up being the case? Is anyone successfully using UE5 on NixOS beyond just seeing if they can?

I got it to work somehow with a nix shell, but I don't see myself using it for production like this. It also seems like some features were not working well because not properly ported to linux and Vulkan on Nvidia GPUs for instance.
It seems it might work better on an Ubuntu based distro though.
If anyone has more news please share them.

**Comment by KingKrouch** [comment]
> Maybe instead packaging something like [Epic Asset Manager](https://github.com/AchetaGames/Epic-Asset-Manager) would be better. I heard the dev mention that with the new binary releases, they were planning on implementing direct downloads and version management in the app

One of the developers of the app is considering adding prefixes and subfixes to launch options for the engine, so that means that one could hypothetically use Steam-run or launch the engine inside of a Distrobox container very easily once that is implemented. I think this would probably be a good approach for running the engine.

**Comment by johanno** [comment]
> > Maybe instead packaging something like [Epic Asset Manager](https://github.com/AchetaGames/Epic-Asset-Manager) would be better. I heard the dev mention that with the new binary releases, they were planning on implementing direct downloads and version management in the app
> 
> Did this end up being the case? Is anyone successfully using UE5 on NixOS beyond just seeing if they can?

I mean I am just looking here for the exact same reason. However I managed to get UE5 running using Lutris on Debian once. So I think this would be a (bad performing) backup plan. It crashes sometimes and is a bit laggy, but overall usable with proton.

**Comment by johanno** [comment]
Ok little update on compiling from sources:

```nix{ pkgs ? import <nixpkgs> {}, ... }:
pkgs.mkShell {
  packages = with pkgs; [
    stdenv.cc.cc
    bash
    curl
  ];
  
  shellHook = ''
     # get unreal source code.
     cd UnrealEngine
     steam-run Setup.sh
     # will fail on curl missing ca file for server.
     # search for curl in files with type shell script in dolphin....
     # probably Engine/Build/BatchFiles/Linux/SetupToolchain.sh // add -k to curl call to skip verification
     steam-run GenerateProjectFiles.sh
     steam-run make
     # once build you can run Engine/Binaries/Linux/UnrealEditor with steam-run
  '';
}
```

I managed to compile it from the sources. I made 5.4.4 working and have been able to load a project and run around in it. 
I don't know if anything else is not working, but it takes forever of course to compile. Well maybe I will now get around learning UE and see what linux is lacking.

**Comment by johanno** [comment]
And in case you want to rebuild your stuff with vs code rename Engine/Build/BatchFiles/Linux/Build.sh and replace it with this:
`#!/usr/bin/env bash
SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
steam-run $SCRIPT_DIR/Build.orig.sh $@`

**Comment by Murazaki** [comment]
> Ok little update on compiling from sources:
> 
> ```
> pkgs.mkShell {
>   packages = with pkgs; [
>     stdenv.cc.cc
>     bash
>     curl
>   ];
>   
>   shellHook = ''
>      # get unreal source code.
>      cd UnrealEngine
>      steam-run Setup.sh
>      # will fail on curl missing ca file for server.
>      # search for curl in files with type shell script in dolphin....
>      # probably Engine/Build/BatchFiles/Linux/SetupToolchain.sh // add -k to curl call to skip verification
>      steam-run GenerateProjectFiles.sh
>      steam-run make
>      # once build you can run Engine/Binaries/Linux/UnrealEditor with steam-run
>   '';
> }
> ```
> 
> I managed to compile it from the sources. I made 5.4.4 working and have been able to load a project and run around in it. I don't know if anything else is not working, but it takes forever of course to compile. Well maybe I will now get around learning UE and see what linux is lacking.

That's great and all but we (NixOS or the community) can't distribute Unreal rebuilt because of the engine license. And I don't like the idea of rebuilding it when it's already built for linux somewhere we can pick with just a download link ? Ofc there's the LD issues but that can be handled.

**Comment by KingKrouch** [comment]
> That's great and all but we (NixOS or the community) can't distribute Unreal rebuilt because of the engine license. And I don't like the idea of rebuilding it when it's already built for linux somewhere we can pick with just a download link ? Ofc there's the LD issues but that can be handled.

I wonder if this can be approached similarly to how CISCO's Packet Tracer is in the Nix package manager. It still technically requires the debian package for it to be downloaded from the website and referenced, and I'd imagine that maybe something similar could be done with Unreal's engine binaries (Although I'd imagine subsequent engine versions, which are regular, would present issues).

**Comment by Murazaki** [comment]
> I wonder if this can be approached similarly to how CISCO's Packet Tracer is in the Nix package manager. It still technically requires the debian package for it to be downloaded from the website and referenced, and I'd imagine that maybe something similar could be done with Unreal's engine binaries (Although I'd imagine subsequent engine versions, which are regular, would present issues).

Is there really a deb package ? For all I know builds are only here behind a login and provided EULA acceptance : https://www.unrealengine.com/en-US/linux

**Comment by KingKrouch** [comment]
> Is there really a deb package ? For all I know builds are only here behind a login and provided EULA acceptance : https://www.unrealengine.com/en-US/linux

Oh, I was referring to packettracer when I was talking about a debian package. I know that UE4 releases for Linux are in ZIP files. Should've clarified.

**Comment by johanno** [comment]
> > Ok little update on compiling from sources:
> > ```
> > pkgs.mkShell {
> >   packages = with pkgs; [
> >     stdenv.cc.cc
> >     bash
> >     curl
> >   ];
> >   
> >   shellHook = ''
> >      # get unreal source code.
> >      cd UnrealEngine
> >      steam-run Setup.sh
> >      # will fail on curl missing ca file for server.
> >      # search for curl in files with type shell script in dolphin....
> >      # probably Engine/Build/BatchFiles/Linux/SetupToolchain.sh // add -k to curl call to skip verification
> >      steam-run GenerateProjectFiles.sh
> >      steam-run make
> >      # once build you can run Engine/Binaries/Linux/UnrealEditor with steam-run
> >   '';
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I managed to compile it from the sources. I made 5.4.4 working and have been able to load a project and run around in it. I don't know if anything else is not working, but it takes forever of course to compile. Well maybe I will now get around learning UE and see what linux is lacking.
> 
> That's great and all but we (NixOS or the community) can't distribute Unreal rebuilt because of the engine license. And I don't like the idea of rebuilding it when it's already built for linux somewhere we can pick with just a download link ? Ofc there's the LD issues but that can be handled.

Well I for my part could not fix the LD issues. I used autopatchelf and fixed all missing dependencies, but it won't launch a project for me. There would be needed some deeper digging. Compiling from sources went relatively smooth. It only takes a few hours or longer depending on your hardware.

**Comment by Murazaki** [comment]
> Is there a reason you don't use the `targetPkgs` variable you defined?

AFAIK, that's how BuildFHSEnv works. Here's a doc :

https://ryantm.github.io/nixpkgs/builders/special/fhs-environments/

**Comment by Murazaki** [comment]
There's was an issue with direnv that loops when using an FHS and recursive nix behavior (must be similar to https://github.com/nix-community/nix-direnv/issues/496). I moved the variables before the FHS to avoid the issue :

```nix
{ pkgs ? import <nixpkgs> {} }: let
  stdenv = pkgs.llvmPackages_18.stdenv;
  dotnetPkg = (with pkgs.dotnetCorePackages; combinePackages [
      sdk_9_0
    ]);
  deps = (with pkgs; [
    zlib
    zlib.dev
    openssl
    dotnetPkg
  ]);
in
(pkgs.buildFHSEnv {
  name = "UnrealEditor";

  targetPkgs = pkgs: (with pkgs;
  [ udev
    alsa-lib
    mono
    dotnet-sdk
    stdenv
    clang_18
    icu
    openssl
    zlib
    SDL2
    SDL2.dev
    SDL2 SDL2_image SDL2_ttf SDL2_mixer
    vulkan-loader
    vulkan-tools
    vulkan-validation-layers
    glib
    libxkbcommon
    nss
    nspr
    atk
    mesa
    dbus
    pango
    cairo
    libpulseaudio
    libGL
    expat
    libdrm
    wayland
  ]) ++ (with pkgs.xorg;
  [ 
    libICE
    libSM
    libX11
    libxcb
    libXcomposite
    libXcursor
    libXdamage
    libXext
    libXfixes
    libXi
    libXrandr
    libXrender
    libXScrnSaver
    libxshmfence
    libXtst
  ]);

  runScript = "zsh";

  NIX_LD_LIBRARY_PATH = pkgs.lib.makeLibraryPath ([
    stdenv
  ] ++ deps);
  NIX_LD = "${stdenv.cc.libc_bin}/bin/ld.so";
  nativeBuildInputs = [ 
  ] ++ deps;

  shellHook = ''
    DOTNET_ROOT="${dotnetPkg}";
    DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1;
  '';
}).env

```

**Comment by ccalhoun1999** [comment]
Has anyone managed to write a flake you can through in an engine directory to build the engine?

**Comment by KingKrouch** [comment]
> Has anyone managed to write a flake you can through in an engine directory to build the engine?

I don't know how helpful this is, but I have mainly been messing around with binary builds of the engine with a custom developer shell, and I had to modify some of the bash scripts (mainly the Setup.sh, and the files in ``Engine/Build/BatchFiles/Linux/``) to use ``#!/usr/bin/env bash`` instead of ``#!/bin/bash``.

I recently submitted a pull request to the Unreal Engine GitHub to hopefully adjust this, so it will work better on NixOS.

**Comment by ccalhoun1999** [comment]
If you are using ue 5.7 in wayland that's probably the source of most of your issues. Supposedly they don't spawn windows with proper window types so wayland struggles (for example you can't drag and drop blueprint nodes).

**Comment by Murazaki** [comment]
I wanted to notify anyone wanting to use Unreal that I found this project, and it looks like it does integrate Unreal well, and with Rider support ?

If anyone else wants to try, maybe we can think about integrating some of its stuff as Unreal helpers in NixOS ?

https://github.com/Adrastie/NixOSUnreal

**Comment by evanlhatch** [comment]
Love this idea. This would be excellent to build on for use as a build server as well as a dev environment 

I don't have a lot of time to work on this stuff but I'd contribute grant money if people are serious

> I wanted to notify anyone wanting to use Unreal that I found this project, and it looks like it does integrate Unreal well, and with Rider support ?
> 
> If anyone else wants to try, maybe we can think about integrating some of its stuff as Unreal helpers in NixOS ?
> 
> https://github.com/Adrastie/NixOSUnreal
